### DISCUSSION TITLE
Welcome to KivyMD Discussions!

### DISCUSSION THREAD

**Discussion by HeaTTheatR** [HeaTTheatR/KivyMD#1077] [Ideas]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by PessahDerek** [comment]
Hello, I'm Derek,  I'm sort of a  beginner in Kivy and kivymd. I did not want to raise to an issue before being sure it was one but I think the MDTopAppBar is not working. I'm getting an error that it is an unknown class. I even copy pasted the code in the docs, same error. Kindly help me out

**Comment by ndonkoHenri** [comment]
Hello. Just to talk of my ideas about the get_rgb() function of the MDColorPicker.
_The problem might seem little, but caused to alot of little bugs in my code's functionalities before I noticed it._

Here is the function:
```
    def get_rgb(self, color: list) -> list:
       """Returns an ''RGB'' list of values from 0 to 255."""

        return [
            int(value * 255)
            for value in (color[:-1] if len(color) == 4 else color)
        ] 
```

This code works perfect when provided a color of the rgba form(length=4).
But when a color instead in rgb form(length=3) is chosen from the Colorpicker and passed as parameter to this function, a list of enormous values( >255) are returned. **This contradicts the function's description(which says values between  0 and 255), and leads to errors if the returned values are used for other stuffs like setting or updating the color of stuffs.**

Here are my ideas:

- [For Clarity] limit the function to only lists of length 4(rgba), and raising a clear error-message if any other length is passed; 
- [For Explicity] state in the function's description, that the color parameter must be of length 4(rgba form), making it obvious to developers

I never wanted to create an issue, since (i thought) this is not really one.
Thanks for reading and responding.....(let's discuss)

**Reply by ndonkoHenri** [reply]
@HeaTTheatR what about this?

**Comment by ndonkoHenri** [comment]
## 1. Idea on the function kivy.utils.get_random_color():

```
def get_random_color(alpha=1.0):
    '''Returns a random color (4 tuple).

    :Parameters:
        `alpha`: float, defaults to 1.0
            If alpha == 'random', a random alpha value is generated.
    '''
    from random import random
    if alpha == 'random':
        return [random(), random(), random(), random()]
    else:
        return [random(), random(), random(), alpha]
```

**My idea is to** add an optional-parameter named '_exclude_' or '_exclude_colors_' which will avoid returning a specific color or a range of colors.

## 2. Why don't I make a PR instead showing the implementation?
The return of the `random()` function is **extremely** unpredictable, making it difficult(for me) to check the presence of the colors to be excluded. (I am still a nooby!)

## 3. Why am I having this idea?
When using this function to obtain a random background color for a component(MDLabel, MD..Button), I noticed that after running the program several times(e.g through hot-reload), the `background color ==  foreground color` making it impossible to distinguish between them! (If I see this happen on any app, I will immidiately delete it!)
Hence, excluding the foreground color range in the `get_random_color` function could be a great idea to resolve this problem.


Thanks for reading. **Let me know how you guys find my idea...**
- Have you ever faced a similar problem when using this function?
- Could my idea be of help?

**Comment by HeaTTheatR** [comment]
@ndonkoHenri No, I didn't have such a problem. Because I didn't use the `get_random_color` function

**Comment by JustmeHello** [comment]
The main page doesnt provide a Button to create a new discussionm how can I write a new discussion?

**Reply by HeaTTheatR** [reply]
![Снимок экрана 2022-12-13 в 22 18 43](https://user-images.githubusercontent.com/16930280/207425124-fa7a5ed5-bd13-4354-91f5-43fc87dde591.png)

**Reply by JustmeHello** [reply]
![2022_12_13_21_22_21_](https://user-images.githubusercontent.com/119158894/207436235-2d772182-1f97-4dce-a01b-92bdf84a7fdd.jpg)

**Reply by JustmeHello** [reply]
> ![Снимок экрана 2022-12-13 в 22 18 43](https://user-images.githubusercontent.com/16930280/207425124-fa7a5ed5-bd13-4354-91f5-43fc87dde591.png)

you can see clearly...i cant start a discussion at my end....what is the problem?

**Reply by HeaTTheatR** [reply]
I don't know...

**Reply by baseplate-admin** [reply]
> you can see clearly...i cant start a discussion at my end....what is the problem?


Same thing is happening to me.

@HeaTTheatR, Could you please see https://github.com/orgs/community/discussions/29#discussioncomment-112122 ?

**Comment by Dreamer-4pda** [comment]
Cant start discussion in Q&A
Ask here.
How i can clear all widgets inside instanse_tab?

**Reply by HeaTTheatR** [reply]
```python
    for instance_tab_label in self.tabs.get_tab_list():
        self.tabs.remove_widget(instance_tab_label)
```

https://kivymd.readthedocs.io/en/1.1.1/components/tabs/#kivymd.uix.tab.tab.MDTabs.get_tab_list

**Comment by Dreamer-4pda** [comment]


**Reply by HeaTTheatR** [reply]
Unfortunately, I don't understand your question...

**Reply by Dreamer-4pda** [reply]
i want get current tab label inside my function

**Comment by felixus97** [comment]
Hello!

Can somebody else take a look at issue #1437 please?

It has been closed, albeit hastily. While the matter may seem like an error on my part at first, I have looked at the related code and understood how they work before posting the issue.

I have written more comments to detail the problem.

Thanks

**Comment by vicem-git** [comment]
Hello, Im experimenting a problem with the theming of my kvmd app. I dont know if this is the right channel, I hope you can help me. I have an app using screen manager, with a couple of screens (im leaving out that part of the code), and as you can see in the attached image, colours are not applying. I tried switching to dark theme and this works, but still color palettes dont work :(

As far as i understand, there is no need to apply theme/colors in each screen/widgets from my kv file, right ? There is no mention of this in docs/ tutos... I did manage to change color of a button from the kv file for example, but i dont know what would be the point of themes if then i need to customize every widget/screen myself from the builder...

i have kivy/ kivymd installed only in the virtual environment of my project, i tried uninstaling-reinstalling but still doesnt fix it.


CODE >>>>

```python
import sqlite3 as sql
from kivymd.app import MDApp
from kivy.lang import Builder
from kivy.metrics import dp
from kivy.uix.screenmanager import ScreenManager, Screen
from kivy.properties import ObjectProperty
from kivymd.uix.datatables import MDDataTable
from kivy.uix.anchorlayout import AnchorLayout


class ViewMan(ScreenManager):

class LoginPage(Screen):

class MainPage(Screen):

class AdhesionPage(Screen):


class AdhApp(MDApp):
    def build(self):
        self.theme_cls.theme_style = "Dark"
        self.theme_cls.primary_palette = "Blue"
        self.theme_cls.accent_palette = "DeepPurple"

        return Builder.load_file("builder.kv")

if __name__ == '__main__':
    AdhApp().run()
```

WITH theme_cls.theme_style = "Light"
![screen_notW](https://user-images.githubusercontent.com/47831722/217780341-2c4f68ef-233d-4c33-b070-6474bd3cf4f7.png)

WITH theme_cls.theme_style = "Dark"
![dark_colors_notW](https://user-images.githubusercontent.com/47831722/217782658-82ee541c-5776-49e4-9246-32bbcd15f24b.png)

**Reply by vicem-git** [reply]
To follow up, I quickly did this simple app with just a button, to show that same thing is happening... colors do not apply !

PY FILE:

`from kivymd.app import MDApp
from kivy.lang import Builder


class MyApp(MDApp):
    def build(self):
        self.theme_cls.primary_palette = "Blue"

        return Builder.load_file("color_theme.kv")


MyApp().run()`


KV FILE: 

`Screen:
    Button:
        text:"button"
        size_hint:(0.5,0.5)`

RESULT:
![sameprob](https://user-images.githubusercontent.com/47831722/217786289-f7ad0bf6-e5e7-4d17-97bd-42c62cd758f2.png)

**Comment by HeaTTheatR** [comment]
@ozymandiasgeist Judging by your screenshots, you are not using any KivyMD widget. Kivy is not KivyMD. Please read the documentation for the KivyMD library - https://kivymd.readthedocs.io/en/latest/getting-started/

**Comment by ukblumf** [comment]
Hi, I'm Neil. New to KivyMD but liking very much what I've been able to do with it so far. However I've hit a snag and I'm unsure on how to proceed. I'm trying to implement a drag and drop of ListItems within a ListView, basically allowing a user to manually sort the ListItems.  
It could be a limitation of the browser rendition and may work fine on mobile/tablets, but it seems I can't get the IconLeftWidget to recognise any of the touch methods.  
I suspect it maybe to do with propagate_touch_to_touchable_widgets method but not clear how to implement it. 

Any help welcome! Thanks.

**Reply by Dreamer-4pda** [reply]
Yep it's a bug, see in issues.
Now working only on_touch function.
Or don't use top bar in your project.

**Comment by AnderSchofield** [comment]
Hey guys! First off, my apologies if this isn't the right channel to bring this up, but I came across an issue while tinkering with MD's code. The MDNavigationRaid module has a function that references a variable that doesn't exist in its docstrings. I tried to edit it on GitHub, but I couldn't find the files. However, I did notice that the latest module's docstrings have fixed this issue. Unfortunately, the website and PDF contents of v.1.1.1 still contain the error.

Here's the issue I found:

def set_radius(self, *args):
    if self.rounded_button:
        self._radius = self.radius = self.height / 4

And here's the fixed code (Already inside the module's docstring as said)

def set_radius(self, *args):
    if self.rounded_button:
        value = self.height / 4
        self.radius = [value, value, value,  #value]
        self._radius = value
Thanks for your help in advance!

**Comment by gijocode** [comment]
Hi Guys,

I was trying to create a `MDSegmentedControl` for my App, when I discovered that the width of the segment panel, `segment_panel_width` is fixed to `"320dp"`. Its not an issue on my laptop, but when I change my window size to phone specifications, my `MDSegmentedControl` is spilling out of the sides.

I see that `segment_panel_height` can be set by the user. Is there a reason the `segment_panel_width` is fixed?

If it can be changed, I will raise a PR for the same, to make the `segment_panel_width` a settable property.

**Reply by HeaTTheatR** [reply]
https://github.com/kivymd/KivyMD#warning
https://git.org.ru/KivyMD/kivymd/issues

**Comment by NightFurySL2001** [comment]
Since I can't open a new discussion, I'll ask here:

In https://kivymd.readthedocs.io/en/1.1.1/components/navigationdrawer/#kivymd.uix.navigationdrawer.navigationdrawer.MDNavigationDrawer.type the demo shows a standard MDNavigationDrawer that opens and pushes the whole screen to the right. However I am unable to reproduce the same result. Is there any sample code for the demonstrated picture shown?

**Comment by ttjuegos** [comment]
hello
How can I clean text in MDTextField on_focus?
or select all text

**Comment by snckmykek** [comment]
Hi, i trying to add MDBottomNavigation into MDNavigationLayout and getting error: The MDNavigationLayout must contain only `MDNavigationDrawer` and `ScreenManager`
But documentation says MDBottomNavigation works like ScreenManager.
If i change kivymd.uix.navigationdrawer in 637 line:
from
`if not isinstance(
    widget, (MDNavigationDrawer, ScreenManager, MDTopAppBar)
):`
to
`if not isinstance(
    widget, (MDNavigationDrawer, ScreenManager, MDTopAppBar, MDBottomNavigation)
):`
its works good. i think u can add it

**Comment by mutiarakhatami** [comment]


**Reply by HeaTTheatR** [reply]
This is not related to the KivyMD library. Contact the Kivy server for help

**Comment by HeaTTheatR** [comment]
Attach a minimal code example

**Comment by HeaTTheatR** [comment]
**If you would like to get an answer to your question quickly, please follow the following small terms:**

If you get an error in your code, please describe your question in one message and attach **MINIMUM CODE **.

**MINIMUM CODE** is:

Code in which there are no unnecessary imports, classes, functions, widgets that are not related to the reproduction of the error.

KV string should be in the code, not in a separate file, for example:

```python
[imports]

KV = """
    ...
"""

class Test(MDApp):
    def build(self):
        return Builder.load_string(KV)


Test().run()
```

If your code exceeds the allowed number of characters in one message, then this is not the **MINIMUM CODE**.


Your code should run in three clicks: **Copy/Paste/Run**
If this cannot be done without leaving `Discord`, this is not a **MINIMUM CODE**.

Before attaching the minimal code, describe how to interact with this code to reproduce the problem. Describe what this code does. Describe what this code does not do as you expect. Describe how you want this code to work.

**Why are these requirements:**

**1.** When you refactor your code in accordance with all of the above, you will find the error yourself 90% of the time.

**2**. You respect the developers who will look at your code. They don't need to look for an error and dig into large code if your code is 50-100 lines long.

**3.** No need to attach screenshots of the code. No need to attach screenshots of logs. Attach text only.

**4.** You don't need to use the https://pastebin.com service to demonstrate your code

**5.** Additional information - https://discord.com/channels/566880874789076992/618108815774056488/739213696680853524

**Thanks for reading!**

**Reply by SebastianObi** [reply]
OK, thanks for the info. I will then create a new post with the proper details.

**Reply by SebastianObi** [reply]
Hello,
I am a newbie in the use of kivy/kivymd.

I have a problem with displaying variable height data in the RecycleView. When each element is the same height, scrolling works fine. But if the elements are different height, then the scrolling is stuttering and jumping. So when new widgets/elements are loaded.

I have already tried to find a solution using other articles on the Internet. I am unsure if it is my error or possibly a bug or both.
I am also unsure if it is a kivymd or kivy problem.

I attached a minimal code example.

Platform: Android
Kivy: 2.2.1
KivyMD: 1.2.0.dev0

Thank you for your help.

[main.txt](https://github.com/kivymd/KivyMD/files/12430016/main.txt)

**Reply by HeaTTheatR** [reply]
You cannot use elements of different heights in RecycleView

**Reply by SebastianObi** [reply]
Thank you for the quick feedback.
I already thought that the RecycleView might not be intended for this. It was just unclear. I will then probably have to work with normal widgets on the page. A RecycleView would just have a much better performance in the display of larger data sets than individual widgets.

**Comment by mutiarakhatami** [comment]
hello, i have problem with pytesseract in kivymd for android.. I'm making an image to text application, how do you do pytesseract so that it works on Android?

    def convert_image_to_text(self, image_path):
        ps.pytesseract.tesseract_cmd = r'C:\Program Files\Tesseract-OCR\tesseract.exe'
        try:
            image = Image.open(image_path)
            text = ps.image_to_string(image)
            self.root.get_screen('gambar').ids.card_text.text = text
        except Exception as e:
            toast('Error:' + str(e))
            
in my code ps.pytesseract.tesseract_cmd = r'C:\Program Files\Tesseract-OCR\tesseract.exe' it can only be run on my laptop, how can I make it run on the pytesseract android? please help me

**Comment by HeaTTheatR** [comment]
@mutiarakhatami Your question is not related to the KivyMD library

**Comment by Mr-robot5** [comment]
where is the mistake? writes: kivy.factory.FactoryException: Unknown class <MDToolbar>
I did everything the same as one blogger two years ago. Or have the standards changed in two years?
swears on MDToolbar

`from kivy.lang import Builder
from kivy.uix.boxlayout import BoxLayout
from kivy.properties import StringProperty, ListProperty

from kivymd.app import MDApp
from kivymd.theming import ThemableBehavior
from kivymd.uix.list import OneLineIconListItem, MDList

KV = '''
# Menu item in the DrawerList list.
<ItemDrawer>:
    theme_text_color: "Custom"
    on_release: self.parent.set_color_item(self)

    IconLeftWidget:
        id: icon
        icon: root.icon
        theme_text_color: "Custom"
        text_color: root.text_color


<ContentNavigationDrawer>:
    orientation: "vertical"
    padding: "8dp"
    spacing: "8dp"

    AnchorLayout:
        anchor_x: "left"
        size_hint_y: None
        height: avatar.height

        Image:
            id: avatar
            size_hint: None, None
            size: "56dp", "56dp"
            source: "data/logo/kivy-icon-256.png"

    MDLabel:
        text: "KivyMD library"
        font_style: "Button"
        size_hint_y: None
        height: self.texture_size[1]

    MDLabel:
        text: "kivydevelopment@gmail.com"
        font_style: "Caption"
        size_hint_y: None
        height: self.texture_size[1]

    ScrollView:

        DrawerList:
            id: md_list



Screen:

    MDNavigationLayout:

        ScreenManager:

            Screen:

                BoxLayout:
                    orientation: 'vertical'

                    MDToolbar:
                        title: "Navigation Drawer"
                        elevation: 10
                        left_action_items: [['menu', lambda x: nav_drawer.set_state("open")]]

                    Widget:


        MDNavigationDrawer:
            id: nav_drawer

            ContentNavigationDrawer:
                id: content_drawer
'''


class ContentNavigationDrawer(BoxLayout):
    pass


class ItemDrawer(OneLineIconListItem):
    icon = StringProperty()
    text_color = ListProperty((0, 0, 0, 1))


class DrawerList(ThemableBehavior, MDList):
    def set_color_item(self, instance_item):
        """Called when tap on a menu item."""

        # Set the color of the icon and text for the menu item.
        for item in self.children:
            if item.text_color == self.theme_cls.primary_color:
                item.text_color = self.theme_cls.text_color
                break
        instance_item.text_color = self.theme_cls.primary_color


class TestNavigationDrawer(MDApp):
    def build(self):
        return Builder.load_string(KV)

    def on_start(self):
        icons_item = {
            "folder": "My files",
            "account-multiple": "Shared with me",
            "star": "Starred",
            "history": "Recent",
            "checkbox-marked": "Shared with me",
            "upload": "Upload",
        }
        for icon_name in icons_item.keys():
            self.root.ids.content_drawer.ids.md_list.add_widget(
                ItemDrawer(icon=icon_name, text=icons_item[icon_name])
            )


TestNavigationDrawer().run()`

**Comment by Marksuuuu** [comment]
Hello, 

I am new to KivyMD and have utilized GitHub to generate an APK file through Action. The APK creation was successful; however, upon attempting to open the app on my Android phone (running Android OS versions 12 and 13), it consistently crashes.

Here are the versions of Kivy and KivyMD that I am currently using:

- Kivy: 2.1.0
- KivyMD: 1.0.2
- Requests: 2.31.0

**Comment by HeaTTheatR** [comment]
See logs...

**Reply by Marksuuuu** [reply]
![image](https://github.com/kivymd/KivyMD/assets/112992662/383c2cc3-c5f4-4259-8bc6-d4cd47ccdfcc)
Here's the logs and there's no error but it keeps crashing

**Comment by HeaTTheatR** [comment]
Use KivyMD from master branch

**Reply by Marksuuuu** [reply]
Thank you!!.

I solved my problem. I found that the .spec file and the requirements section need to have all small caps, and osx.kivy_version needs to be the same as in requirements.

**Comment by Maverik37** [comment]
Hello ! 

For an application i want to have MDdatepicker that open only for selecting month , is it possible ?

**Comment by mouse3mic3** [comment]
Greetings! Not sure if this is the correct place for questions as I was unable to start a new discussion. 

I was wondering if MDDataTables will be supported in KivyMD 2 and forward? I am using the latest from master branch and do not find datatables in uix. 

I have an old project which relies on it and I work around this by copying the datatables' py and kv files from older version. It seems to work mostly fine, I had to adjust some theming cls attributes in the datatables.kv file due to API breaks.

**Reply by T-Dynamos** [reply]
`MDDataTables` will be moved to [`kivymd_extensions`](https://github.com/kivymd-extensions) in future.

**Comment by moneychin** [comment]
Hello all, I have a question about how to show the Chinese word in KivyMD. I value your expertise and would be grateful for your guidance. Thank you. 

I already have the LabelBase.register(name="DroidSans", fn_regular="cc.ttf"). Do you know how I can make it work? Thank you.  

Here is the code:

from kivymd.app import MDApp
from kivy.lang import Builder
from kivy.core.window import Window
from kivy.uix.screenmanager import ScreenManager, Screen
from kivy.core.text import LabelBase
from kivymd.theming import ThemeManager
from kivy.properties import ObjectProperty
LabelBase.register(name="DroidSans", fn_regular="cc.ttf")


class TestApp(Screen):
    pass


class SecondLayout(Screen, ThemeManager):
    manager = ObjectProperty()
    nav_drawer = ObjectProperty()


class LayoutManager(ScreenManager):
    pass


class MyApp(MDApp):
    def build(self):
        Window.size = (360, 640)
        return Builder.load_file("first.kv")


if __name__ == "__main__":
    MyApp().run()

Here is the KivyMD file: 

LayoutManager:
    TestApp:
    SecondLayout:

<TestApp>:
    name: "login"
    MDScreen:
        md_bg_color: self.theme_cls.secondaryContainerColor
        MDCard:
            size_hint: None, None
            size: 800, 1000
            pos_hint: {"center_x": .5, "center_y": .5 }
            padding: 80
            spacing: 18
            orientation: "vertical"

            MDLabel:
                id: welcome_label
                text:'哈哈'
                font_name: "DroidSans"
                font_style: "Headline"
                halign: 'center'
